commit a0a104ceb3720c15051ba4cc27192d5c301e8097
Author:     Isaac Yang <isaacy@nvidia.com>
AuthorDate: Thu May 28 15:43:37 2026 -0700
Commit:     GitHub <noreply@github.com>
CommitDate: Thu May 28 22:43:37 2026 +0000

    [2.8]: cherry-pick PR 4684 and 4704 Improve security by loading classes in allow_list only (#4701)
    
    Address comments
    
    Switch to block_list
    
    Update doc and improve coverage of unit tests
    
    Address comments
    
    Add public API
    
    Fix format issue
    
    Revert back to allow_list
    
    Fix one unit test
    
    Address new comments
    
    Improve more
    
    Address more comments
    
    Update the class_allow_list
    
    Address new comments
    
    Add cache to speed up
    
    Update document for more clarities
    ### Description
    
    Cherry-pick PR#4684 and PR#4704
    
    ### Types of changes
    <!--- Put an `x` in all the boxes that apply, and remove the not
    applicable items -->
    - [ ] Non-breaking change (fix or new feature that would not break
    existing functionality).
    - [x] Breaking change (fix or new feature that would cause existing
    functionality to change).
    - [x] New tests added to cover the changes.
    - [x] Quick tests passed locally by running `./runtest.sh`.
    - [x] In-line docstrings updated.
    - [x] Documentation updated.
---
 nvflare/private/fed/server/server_engine.py | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/nvflare/private/fed/server/server_engine.py b/nvflare/private/fed/server/server_engine.py
index e0003f3d..56011c32 100644
--- a/nvflare/private/fed/server/server_engine.py
+++ b/nvflare/private/fed/server/server_engine.py
@@ -538,6 +538,11 @@ class ServerEngine(ServerEngineInternalSpec, StreamableEngine):
         self.conf = conf
 
     def build_component(self, config_dict):
+        if not self.conf:
+            raise RuntimeError("No configurator set up.")
+        has_authorizer = getattr(self.conf, "has_component_build_authorizer", None)
+        if not callable(has_authorizer) or not has_authorizer():
+            raise RuntimeError("No component build authorizer set up.")
         return self.conf.build_component(config_dict)
 
     def new_context(self) -> FLContext:
